safehttp/plugins/xsrf/xsrf.go

```diff
@@ -34,24 +42,46 @@ type UserIdentifier interface {
 }
 
 // Interceptor implements XSRF protection. It requires an application key and a
-// storage service. The appKey uniquely identifies each registered service and
+// storage service. The AppKey uniquely identifies each registered service and
 // should have high entropy. The storage service supports retrieving ID's of the
-// application's users. Both the appKey and user ID are used in the XSRF
-// token generation algorithm.
+// application's users. Both the AppKey and user ID are used in the XSRF
+// token generation and verification algorithm.
 type Interceptor struct {
 	AppKey     string
 	Identifier UserIdentifier
 }
 
+type tokenCtxKey struct{}
+
+// Token extracts the XSRF token from the incoming request. If it is not
+// present, it returns a non-nil error.
+func Token(r *safehttp.IncomingRequest) (string, error) {
+	tok := r.Context().Value(tokenCtxKey{})
+	if tok == nil {
+		return "", errors.New("xsrf token not found")
+	}
+	return tok.(string), nil
+}
+
 // Before should be executed before directing the safehttp.IncomingRequest to
-// the handler to ensure it is not part of the Cross Site Request
-// Forgery. It checks for the presence of an xsrf-token in the request body and
-// validates it based on the userID associated with the request.
-func (p *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
-	userID, err := p.Identifier.UserID(r)
+// the handler to ensure it is not part of the Cross-Site Request
+// Forgery. In case of state changing requests (all except GET, HEAD and OPTIONS
+// requests), it checks for the presence of an xsrf-token in the request body
+// and validates it based on the userID associated with the request. For all
+// requests, it adds a cryptographically safe XSRF token to the
+// safehttp.IncomingRequest context.
+func (i *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	userID, err := i.Identifier.UserID(r)
 	if err != nil {
 		return w.ClientError(safehttp.StatusUnauthorized)
 	}
```

Comment:
Alternatively you can (and probably should) do:

```
needsValidation := !statePreservingMethods[r.Method]

if needsValidation {
  // do the token extraction and validation
}

tok = xsrftoken.Generate(i.AppKey, userID, r.URL.String())
r.SetContext(context.WithValue(r.Context(), tokenCtxKey{}, tok))
return safehttp.Result{}
```

Reply:
Initially, that's the way it was (and I prefer it as well because I don't need to do token generation twice), but Rob suggested moving token verification outside the if-clause so I don't have most of the function logic indented. Which one do you prefer? @kele @empijei

Reply:
I don't have a strong preference. You can also have a `Interceptor.extractToken()` method and have less lines in this one.

Reply:
factoring it out in a function works for me, otherwise I prefer to have it unindented.